clients/database/request_metric.go

```diff
@@ -0,0 +1,47 @@
+package database
+
+import (
+	"context"
+	"time"
+
+	"github.com/uptrace/bun"
+)
+
+// ProxiedRequestMetric contains request metrics for
+// a single request proxied by the proxy service
+type ProxiedRequestMetric struct {
+	bun.BaseModel `bun:"table:proxied_request_metrics,alias:prm"`
+
+	ID                          int64 `bun:",pk,autoincrement"`
+	MethodName                  string
+	BlockNumber                 *int64
+	ResponseLatencyMilliseconds int64
+	RequestTime                 time.Time
+}
+
+// Save saves the current ProxiedRequestMetric to
+// the database, returning error (if any)
+func (prm *ProxiedRequestMetric) Save(ctx context.Context, db *bun.DB) error {
+	_, err := db.NewInsert().Model(prm).Exec(ctx)
+
+	return err
+}
+
+// ListProxiedRequestMetricsWithPagination returns a page of max
+// `limit` ProxiedRequestMetrics from the offset specified by`cursor`
+// error (if any) along with a cursor to use to fetch the next page
+// if the cursor is 0 no more pages exists
+func ListProxiedRequestMetricsWithPagination(ctx context.Context, db *bun.DB, cursor int64, limit int) ([]ProxiedRequestMetric, int64, error) {
+	var proxiedRequestMetrics []ProxiedRequestMetric
+	var nextCursor int64
+
+	count, err := db.NewSelect().Model(&proxiedRequestMetrics).Where("ID > ?", cursor).Limit(limit).ScanAndCount(ctx)
```

Comment:
I like the cursor approach over offsets 👍.  However, performance isn't always predictable or fast with counts due to how MVCC is implemented in postgres.

Something like:

```go
var proxiedRequestMetrics []ProxiedRequestMetric
err := b.NewSelect().Model(&proxiedRequestMetrics).Where("id > ?", cursor).Order("id desc").Limit(limit).Scan(ctx)

nextCursor := cursor
if len(proxiedRequestMetrics) > 0 {
   nextCursor = proxiedRequestMetrics[len(proxiedRequestMetrics)-1].ID
}

return proxiedRequestMetrics, nextCursor, err
```
would give more consistent performance and eliminate the table scan penalty on count when the index isn't used.

I also modified the cursor logic and added an order by in the example.

For the cursor, without a count we only know the last page if the result set contains less elements than the limit, so we can't return 0 for those instances where there is an exact limit alignment.  In addition, if clients get a 0 cursor and are interested in more results (likely the last page is already stale by the time the client gets the response anyway), they will have to use the result set previously to get the cursor which also breaks encapsulation of the cursor value.

Instead, we always return the next cursor.  If clients reach an empty page, or get num rows less than the limit they quested, then they can choose to stop or continue to poll the cursor.

If they keep polling, whether or not there are enough new inserts to fill the limit, using the nextCursor this way will always result in non-duplicated data and allow clients to stay in sync.  If no results are returned, nextCursor stays the same.  If some results are returned, then nextCursor always is set the the last element so the next call will skip the already returned results.

For the order by addition, result order in postgres is not deterministic unless a sort order is specified.  Therefore without it, the results may contain duplicates, may not return the same results between two calls with the same cursor, and it doesn't guarantee the entire table can be paged since results may be skipped.

Reply:
I agree with all these points on a point by point basis, but conceptually 
- this method is only used by the e2e tests so performance is not the top priority
- the service itself isn't likely to need to read data from this table (caching and serving of cache responses will flow through redis)
- longer term analytics on this data will likely be some form of non latency sensitive export pipeline to a data warehouse that is optimized for running the queries for biz and product questions, and in the meantime the queries will have to be run manually by a dev to get those answers in which case we'll want to write super optimized sql for doing that (either for running directly in the psql console or as part of a golang cli) - but in either case we won't be using this current `ListProxiedRequestMetricsWithPagination` method